server/configuration-validation.js

```diff
@@ -1,125 +1,192 @@
+        indexName === ''
+      ) {
+        log.error(
+          `the "indexName" option of "${name}" should be a non-empty string`
+        )
+      } else if (indexName !== undefined) {
+        options.indexName = indexName
+      }
+
+      return this
+    },
+    validateTransformEntry() {
+      // transformEntry is either undefined or a function
+      if (
+        transformEntry !== undefined &&
+        typeof transformEntry !== 'function'
+      ) {
+        log.error(
+          `the "transformEntry" option of "${name}" should be a function`
+        )
+      } else if (transformEntry !== undefined) {
+        options.transformEntry = transformEntry
+      }
+
+      return this
+    },
+    validateFilterEntry() {
+      // filterEntry is either undefined or a function
+      if (filterEntry !== undefined && typeof filterEntry !== 'function') {
+        log.error(`the "filterEntry" option of "${name}" should be a function`)
+      } else if (filterEntry !== undefined) {
+        options.filterEntry = filterEntry
+      }
 
-  for (const collection in collections) {
-    config[collection] = validateCollectionConfiguration({
-      configuration: collections[collection],
-      collection: collection,
-    })
+      return this
+    },
+    validateMeilisearchSettings() {
+      // Settings is either undefined or an object
+      if (settings !== undefined && !isObject(settings)) {
+        log.error(`the "settings" option of "${name}" should be an object`)
+      } else if (settings !== undefined) {
+        options.settings = settings
+      }
+
+      return this
+    },
+    validatePopulateEntryRule() {
+      // PopulateEntry is either undefined or an object/array/string`
+      if (
+        populateEntryRule !== undefined &&
+        !isObject(populateEntryRule) &&
+        !Array.isArray(populateEntryRule) &&
+        typeof populateEntryRule !== 'string'
+      ) {
+        log.error(
+          `the "populateEntryRule" option of "${name}" should be an object/array/string`
+        )
+      } else if (populateEntryRule !== undefined) {
+        options.populateEntryRule = populateEntryRule
+      }
+
+      return this
+    },
+    validateNoInvalidKeys() {
+      // Keys that should not be present in the configuration
+      Object.keys(excedent).map(key => {
+        log.warn(`The attribute "${key}" of "${name}" is not a known option`)
+      })
+
+      return this
+    },
+    get() {
+      return options
+    },
   }
 }
 
-function validateCollectionConfiguration({ configuration, collection }) {
-  const validApiFields = [
-    'indexName',
-    'transformEntry',
-    'settings',
-    'filterEntry',
-    'populateEntryRule',
-  ]
+function PluginOptions(configuration) {
+  const log = strapi.log // has to be inside a scope
+  const { apiKey, host, ...collections } = configuration
+  const options = {}
 
-  if (configuration === undefined) {
-    return
-  }
+  return {
+    validateConfiguration() {
```

Comment:
You could rename this `validateConfigurationObject`, I think it also adds some clarity if you don't know the file.

Reply:
See thiis comment